doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+
+Stash migration ownership is currently tangled between the adapter crate and the stash crate,
+see [stash: Untangle the stash config collection](https://github.com/MaterializeInc/materialize/issues/20922).
+Ideally the adapter would be in charge of how and why to migrate the catalog contents while the
+stash provides the primitives needed to migrate a stash collection.
+
+Once the migrations are de-tangled, at a high level they will take the following steps:
+
+1. Use `impl From<objects_v(X - 1)::(Object)> for objects_v(X)::(Object)` to convert every object
+   from their old format into their new format.
+2. Use `mz_adapter::catalog::storage::Transaction` to perform any deletions, insertions, or updates
+   that require read/write semantics instead of `mz_stash::transaction::Transaction`.
+
+As a bonus, these migrations start to look very similar to the
+existing [catalog migrations](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/migrate.rs).
+We might be able to combine them into a single logical step instead of having two separate migration
+frameworks (not counting the builtin migration framework).
+
+## Alternatives
+
+- Use multiple shards for the persist implementation
+  once [multi-shard txns](https://github.com/MaterializeInc/materialize/pull/20954) is merged and
+  stable.
+- An alternative approach is to put a trait directly in front of the stash, and implement that trait
+  using the current CRDB implementation and implement the trait using persist.
+    - Pros:
+        - Re-using the existing stash migration might be easier.
+        - We can re-use the existing stash debug tool.
+        - The controller can use persist for durable state for free.
+    - Cons:
+        - Maintaining a generic durable TVC store adds a lot of complexity to the implementation.
+          Much of that complexity is already handled by persist, which is a generic durable TVC
+          store with a dedicated team. Overall, I think it might be a net-win if we can at some
+          point delete the stash crate because it removes a layer of complexity from the adapter.
+        - It's very difficult to get the stash trait right and maintain it. We previously had a
+          stash trait but removed it because it was too hard to maintain.
+- The persist implementation could not use an in-memory cache, and instead go to persist for every
+  read via a read handle. This has better memory usage, but slower reads. The reason that I decided
+  on the current proposed implementation is that the Coordinator already reads the entire durable
+  catalog state into memory for every catalog transaction. So we already need that amount of memory
+  free at any time in order to be able to serve DDL. Certain append only object types, like audit
+  logs and storage usage, don't need to be cached in memory.
+
+## Open questions
+
+- Should the `DurableCatalogTrait` be broken up into two traits? A read only trait and a read/write
```

Comment:
what do you think might be benefits of this?

Reply:
One vision for pv2 is to have a read only catalog API that `adapterd`/`pgwired`/`whatever-we're-calling-thisd` can directly query without going through the coordinator. The Coordinator will use a read/write API to make changes for DDL. See https://www.notion.so/materialize/Use-case-Isolation-Milestone-1-72db9064b01b43fdafd0a807a28c7342#5e167a822d764e75b9957d89fbb31c4d. So splitting this out into two traits that look like

```Rust
pub trait ReadOnlyDurableCatalogState {
    ...
}

pub trait DurableCatalogState: ReadOnlyDurableCatalogState {
    ...
}
```

might help prepare for that. It also might be a bit premature and it will be easy enough to split them out later.

Reply:
I think this is a great idea, the `ReadOnlyDurableCatalogState` could also be used for debug tools so we can statically guarantee we're not accidentally modifying the Catalog

Reply:
if you want inspiration for this, old persist did exactly this sort of thing for the Blob and Consensus traits https://github.com/MaterializeInc/materialize/pull/9574